libs/optimizer/src/lib/optimizer.ts

```diff
@@ -0,0 +1,348 @@
+   * indices of solutions points in
+   */
+  solutionIndices: number[];
+  /**
+   * the result is optimal (no need to get a next result of Iterator<OptimizationResult, OptimizationResult>)
+   */
+  optimal: boolean;
+}
+
+const ZERO_SCORE: OptimizationResult = {
+  score: 0,
+  lengthKm: 0,
+  multiplier: 0,
+  solutionIndices: [],
+  optimal: true,
+};
+
+/**
+ * returns an iterative optimizer that computes iteratively the score for the flight. At each iteration, the score
+ * should be a better solutions.
+ * @param request the OptimizationRequest. if request.options is undefined, then there will be one iteration, and the result
+ *                will be the best solution
+ * @param scoringRules the ScoringRules to apply for computation
+ * @return an Iterator over the successive OptimizationResult
+ * @see README.md
+ */
+export function* getOptimizer(
+  request: OptimizationRequest,
+  scoringRules: ScoringRules,
+): Iterator<OptimizationResult, OptimizationResult> {
+  if (request.track.points.length == 0) {
+    console.warn('Empty track received in optimization request. Returns a 0 score');
+    return ZERO_SCORE;
+  }
+  const originalTrack = request.track;
+  const solverTrack = buildValidTrackForSolver(originalTrack);
+  const flight = toIgcFile(solverTrack);
+  const solverScoringRules = toScoringRules(scoringRules);
+  const options = toOptions(request.options);
+  const solutionIterator = solver(flight, solverScoringRules || {}, options);
+  while (true) {
+    const solution = solutionIterator.next();
+    if (solution.done) {
+      console.debug('solution', JSON.stringify(solution.value, undefined, 2));
+      return toResult(solution.value, originalTrack, solverTrack);
+    }
+    yield toResult(solution.value, originalTrack, solverTrack);
+  }
+}
+
+/**
+ * the solver requires at least 5 points, so if there is not enough points,
+ * we create points between existing ones
+ */
+function buildValidTrackForSolver(track: ScoringTrack) {
+  if (track.points.length >= MIN_POINTS) {
+    return track;
+  }
+  console.debug(`not enough points (${track.points.length}) in track. Interpolate intermediate points`);
+  // avoid potential future bugs with a defensive copy to avoid unwanted mutations on inputs.
+  let newTrack: ScoringTrack = deepCopy(track);
+  while (newTrack.points.length < MIN_POINTS) {
+    const segments: ScoringTrack[] = [];
+    for (let i = 1; i < newTrack.points.length; i++) {
+      // split each segment of the track into two segments
+      segments.push(
+        createSegments(
+          newTrack.points[i - 1],
+          newTrack.points[i],
+          track.startTimeSec,
+          NUM_SEGMENTS_BETWEEN_POINTS,
+          DISTRIBUTION_FACTOR_FOR_ADDED_POINTS,
+        ),
+      );
+    }
+    newTrack = concatTracks(...segments);
+  }
+  console.debug(`new track has ${newTrack.points.length} points`);
+  return newTrack;
+}
+
+
+/**
+ * build a fake igc file from a track, so that the solver can use it
```

Comment:
"Create an IGC file"

"so that the solver can use it" -> I prefer not to comment on a particular usage but rather than what the function does. Usages can change.

Reply:
Ok. Changed